pkg/operator/csr/cert_controller.go

```diff
@@ -0,0 +1,322 @@
+// creates a client certificate and rotates it before it becomes expired by using csrs. The client
+// certificate generated is stored in a specific secret with the keys below:
+// 1). tls.key: tls key file
+// 2). tls.crt: tls cert file
+type clientCertificateController struct {
+	ClientCertOption
+	CSROption
+
+	hubCSRLister    certificateslisters.CertificateSigningRequestLister
+	hubCSRClient    csrclient.CertificateSigningRequestInterface
+	spokeCoreClient corev1client.CoreV1Interface
+	controllerName  string
+
+	// csrName is the name of csr created by controller and waiting for approval.
+	csrName string
+
+	// keyData is the private key data used to created a csr
+	// csrName and keyData store the internal state of the controller. They are set after controller creates a new csr
+	// and cleared once the csr is approved and processed by controller. There are 4 combination of their values:
+	//   1. csrName empty, keyData empty: means we aren't trying to create a new client cert, our current one is valid
+	//   2. csrName set, keyData empty: there was bug
+	//   3. csrName set, keyData set: we are waiting for a new cert to be signed.
+	//   4. csrName empty, keydata set: the CSR failed to create, this shouldn't happen, it's a bug.
+	keyData []byte
+}
+
+// NewClientCertificateController return an instance of clientCertificateController
+func NewClientCertificateController(
+	clientCertOption ClientCertOption,
+	csrOption CSROption,
+	hubCSRInformer certificatesinformers.CertificateSigningRequestInformer,
+	hubCSRClient csrclient.CertificateSigningRequestInterface,
+	spokeSecretInformer corev1informers.SecretInformer,
+	spokeCoreClient corev1client.CoreV1Interface,
+	recorder events.Recorder,
+	controllerName string,
+) factory.Controller {
+	c := clientCertificateController{
+		ClientCertOption: clientCertOption,
+		CSROption:        csrOption,
+		hubCSRLister:     hubCSRInformer.Lister(),
+		hubCSRClient:     hubCSRClient,
+		spokeCoreClient:  spokeCoreClient,
+		controllerName:   controllerName,
+	}
+
+	return factory.New().
+		WithFilteredEventsInformersQueueKeyFunc(func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		}, func(obj interface{}) bool {
+			accessor, err := meta.Accessor(obj)
+			if err != nil {
+				return false
+			}
+			// only enqueue a specific secret
+			if accessor.GetNamespace() == c.SecretNamespace && accessor.GetName() == c.SecretName {
+				return true
+			}
+			return false
+		}, spokeSecretInformer.Informer()).
+		WithFilteredEventsInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, c.EventFilterFunc, hubCSRInformer.Informer()).
+		WithSync(c.sync).
+		ResyncEvery(ControllerResyncInterval).
+		ToController(controllerName, recorder)
+}
+
+func (c *clientCertificateController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
```

Comment:
this should be reporting "Progressing" from the moment a new CSR is requested till the certificate is finally pushed into the secret

Reply:
> this should be reporting "Progressing" from the moment a new CSR is requested till the certificate is finally pushed into the secret

that's a good refinement.